### PR TITLE
Implement Color picker pick with custom logic to workaround Godot bug

### LINKED ISSUE #2803: Godot bug: Colour picker doesn't work when window is not default sized
I think we should remove this color picker: 
image
It's broken and useless

### PR DESCRIPTION
**Pull request by 84634E1A607A**
**Brief Description of What This PR Does**
Rewrote color picker's pick button function. Temp fix for Godot issue https://github.com/godotengine/godot/issues/57343

**Related Issues**
Fixes #2803

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

File: src/gui_common/TweakedColourPicker.cs [hunk 1]
```diff
@@ -222,6 +227,15 @@ public override void _Ready()
         presetsContainer = GetNode<GridContainer>("PresetContainer");
         addPresetButton = GetNode<TextureButton>("PresetButtonContainer/AddPresetButton");
 
+        // Picker button logic rewrite #3055
+        // TODO: Revert this PR once https://github.com/godotengine/godot/issues/57343 is solved.
+        baseControl = GetChild(1);
+        var customPickerButton = new ToolButton { Icon = pickerButton.Icon };
+        baseControl.RemoveChild(pickerButton);
```
Comment (hhyyrylainen) [inline src/gui_common/TweakedColourPicker.cs:234]: Is it possible to overwrite the default connection (so unconnect first) and then just link us to the default button? That would mean we mess with Godot's internals less here.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
I experienced the same kind of major lag as @RacerBG when I increased the window size a lot. I think the read from the rendered image starts taking up too long and causes a long queue of input events to process. To solve this, I think we should just sample the colour under the mouse in _Process every 100-200 milliseconds so that we aren't reading so much data from the GPU and converting that over to data readable from C#.

File: src/gui_common/TweakedColourPicker.cs (hunk 1 above)
Comment (hhyyrylainen) [reply src/gui_common/TweakedColourPicker.cs:234]: I guess we don't know the connected object and method name so we can't actually call `Disconnect` here.

File: src/gui_common/TweakedColourPicker.cs [hunk 2]
```diff
@@ -42,6 +42,9 @@ public class TweakedColourPicker : ColorPicker
     private bool rawButtonEnabled = true;
     private bool presetsEditable = true;
     private bool presetsVisible = true;
+    private bool pickingColor;
+    private Color colorBeforePicking;
+    private float pickerTimeElapsed;
     private PickerMode mode;
```
Comment (hhyyrylainen) [inline src/gui_common/TweakedColourPicker.cs:49]:
```suggestion
    private bool presetsVisible = true;
    private PickerMode mode;

    private bool pickingColor;
    private Color colorBeforePicking;
    private float pickerTimeElapsed;

```

File: src/gui_common/TweakedColourPicker.cs [hunk 3]
```diff
@@ -452,6 +500,32 @@ private void OnHtmlColourEditFocusExited()
...
+    {
+        colorBeforePicking = Color;
+        pickingColor = true;
+    }
+
+    private void HandleActiveColourPicking(float delta)
+    {
+        pickerTimeElapsed += delta;
+
+        if (pickerTimeElapsed < 0.2)
```
Comment (hhyyrylainen) [inline src/gui_common/TweakedColourPicker.cs:513]:
```suggestion
        if (pickerTimeElapsed < 0.2f)
```

**Review by hhyyrylainen — APPROVED** [review]
Seems to perform pretty well even at 4K now, which should be good enough.

**Review by Kasterisk — APPROVED** [review]

**Review by RacerBG — COMMENTED** [review]
The lag is gone. Now it is around 1 FPS compared 50-70 FPS reduction in my first test.

But I noticed another problem: Now when I stop the cursor movement, while I'm still in color picker mode, the "cross" (the horizontal and vertical lines) continue to loop over the colors. Also while moving the cursor, the cross lags behind it even though the game is not lagging.

**Comment by hhyyrylainen** [comment]
> But I noticed another problem: Now when I stop the cursor movement, while I'm still in color picker mode, the "cross" (the horizontal and vertical lines) continue to loop over the colors. Also while moving the cursor, the cross lags behind it even though the game is not lagging.

Isn't that just how it works anyway? If you hover over the colour picker itself it ends up updating itself and then reading the updated value causing weirdness.
The mouse lag is an unfortunate side effect of only updating 4 times per second. I suppose we could read the screen texture a few times a second and read from that whenever we want (which should also fix the lag), but that would be harder to implement and I don't think we need to spend the time to code that.

**Comment by hhyyrylainen** [comment]
I'll merge this because this makes the situation much better (and I think fixes a crash). The colours going crazy is a problem with Godot's builtin colour picker as well. If someone really wants the last part of my previous comment can be done by a separate PR.

**Comment by RacerBG** [comment]
> > But I noticed another problem: Now when I stop the cursor movement, while I'm still in color picker mode, the "cross" (the horizontal and vertical lines) continue to loop over the colors. Also while moving the cursor, the cross lags behind it even though the game is not lagging.
> 
> Isn't that just how it works anyway? If you hover over the colour picker itself it ends up updating itself and then reading the updated value causing weirdness. The mouse lag is an unfortunate side effect of only updating 4 times per second. I suppose we could read the screen texture a few times a second and read from that whenever we want (which should also fix the lag), but that would be harder to implement and I don't think we need to spend the time to code that.

I don't know if we are talking about the same thing. I did this:

1. Switched in to color picker mode.
2. Moved the cursor over the colors window or square in a fast manner.
3. Stopped the movement while still the cursor was in the boundaries of the colors window.
4. The cross started to loop over the current cursor x, y position and the last cursor x, y position endlessly.

**Comment by hhyyrylainen** [comment]
That "oscillation" is probably caused by the colour being updated only like 5 times per second so with the fast movement it sees the previous colour, changes to that, and then the new colour ends up being sampled as the previous data was still in use, or something. The move to storing the window render buffer and updating the picked colour from it immediately should fix that, but as I said, I don't think this is a big enough problem to warrant a lot of effort to fix. The oscillation stops if you move the cursor over another element than the colour picker, right?

**Comment by RacerBG** [comment]
> That "oscillation" is probably caused by the colour being updated only like 5 times per second so with the fast movement it sees the previous colour, changes to that, and then the new colour ends up being sampled as the previous data was still in use, or something. The move to storing the window render buffer and updating the picked colour from it immediately should fix that, but as I said, I don't think this is a big enough problem to warrant a lot of effort to fix. The oscillation stops if you move the cursor over another element than the colour picker, right?

Right. And the same behaviour doesn't happen when I try to select a color from the cell surroundings for example.

**Pull request merged**